src/pcm.cpp

```diff
@@ -1,68 +1,99 @@
+		ret.size = utils::Size::make(ret.samples);
+		return ret;
+	}
+}
 
-	drwav m_wav;
-	std::optional<Error> m_error;
-};
+template <>
+Result<PCM> decode<FileFormat::eMp3>(std::span<std::byte const> bytes) {
+	drmp3_config config;
+	drwav_uint64 pcmFrameCount;
 
-std::vector<std::byte> fileBytes(std::string const& path) {
-	static_assert(sizeof(char) == sizeof(std::byte) && alignof(char) == alignof(std::byte));
-	if (auto file = std::ifstream(path, std::ios::binary | std::ios::ate)) {
-		file.unsetf(std::ios::skipws);
-		auto const size = file.tellg();
-		auto buf = std::vector<std::byte>((std::size_t)size);
-		file.seekg(0, std::ios::beg);
-		file.read(reinterpret_cast<char*>(buf.data()), (std::streamsize)size);
-		return buf;
+	auto pData = drmp3_open_memory_and_read_pcm_frames_s16(bytes.data(), bytes.size(), &config, &pcmFrameCount, nullptr);
+	if (pData == nullptr) {
+		return Error::eInvalidData;
+	} else {
+		PCM ret;
+		ret.samples.resize(pcmFrameCount * config.channels);
+		std::copy(pData, pData + (pcmFrameCount * config.channels), ret.samples.begin());
+		ret.meta.rate = config.sampleRate;
+		ret.meta.channels = config.channels;
+		ret.meta.format = config.channels == 2 ? SampleFormat::eStereo16 : SampleFormat::eMono16;
+		ret.size = utils::Size::make(ret.samples);
+		return ret;
 	}
-	return {};
 }
+
 } // namespace
 
-Result<PCM> PCM::fromFile(std::string const& path) {
-	auto const bytes = fileBytes(path);
+Result<PCM> PCM::fromFile(std::string const& path, FileFormat format) {
+	auto const bytes = utils::fileBytes(path);
 	if (bytes.empty()) { return Error::eIOError; }
-	return fromMemory(bytes);
+	return PCM::fromMemory(bytes, format);
 }
 
-Result<PCM> PCM::fromMemory(std::span<std::byte const> wavBytes) {
-	if (wavBytes.empty()) { return Error::eIOError; }
-	WAV wav(wavBytes); // can't use Result pattern here because an initialized drwav object contains and uses a pointer to its own address
-	if (wav.m_error) {
-		return *wav.m_error;
-	} else if (wav.m_wav.channels < 1 || wav.m_wav.channels > 2) {
-		return Error::eUnsupportedChannels;
-	} else {
-		auto const frames = wav.m_wav.totalPCMFrameCount;
-		PCM ret;
-		ret.samples.resize(frames * wav.m_wav.channels);
-		auto const read = wav.read(ret.samples);
-		if (read < frames) { return Error::eUnexpectedEOF; }
-		ret.meta.format = wav.m_wav.channels == 2 ? SampleFormat::eStereo16 : SampleFormat::eMono16;
-		ret.meta.rate = static_cast<std::size_t>(wav.m_wav.sampleRate);
-		ret.meta.channels = static_cast<std::uint8_t>(wav.m_wav.channels);
-		ret.size = utils::Size::make(ret.samples);
-		return ret;
+Result<PCM> PCM::fromMemory(std::span<std::byte const> bytes, FileFormat format) {
+	if (bytes.empty()) { return Error::eIOError; }
+
+	static_assert(static_cast<std::size_t>(FileFormat::eCOUNT_) == 3, "Unhandled file format");
+	switch (format) {
+	case FileFormat::eWav: return decode<FileFormat::eWav>(bytes);
```

Comment:
None of these handles are closed anymore?

Reply:
Oops nope, that should happen inside decode, my bad